server/pkg/apigw/filter/postfilter.go

```diff
@@ -274,16 +286,23 @@ func (j jsonResponse) Handler() types.HandlerFunc {
 			return
 		}
 
-		rw.Header().Add("Content-Type", "application/json")
+		for h, v := range j.params.Header {
+			for _, vv := range v {
+				rw.Header().Add(h, vv)
 
-		switch v := evald.(type) {
-		case string:
-			rw.Write([]byte(v))
-		default:
-			e := json.NewEncoder(rw)
-			err = e.Encode(v)
+				if h == "Content-Type" || h == "content-type" {
+					hasJsonHeader = vv == "application/json"
+				}
+			}
 		}
 
+		if reflect.ValueOf(evald).Kind() != reflect.String && hasJsonHeader {
+			err = (json.NewEncoder(rw)).Encode(expr.UntypedValue(evald))
```

Comment:
Have you tested if this works fine with records, modules, and similar types? We had (have) some issues with json encoding those

Reply:
Tested yes, thanks for the observation